deeppavlov/models/ner/NER_model.py

```diff
@@ -0,0 +1,307 @@
+                                   kernel_initializer=xavier_initializer())
+            self.logits = tf.reshape(pred, [-1, ntime_steps, n_tags])
+
+            log_likelihood, self.transition_params = tf.contrib.crf.crf_log_likelihood(self.logits,
+                                                                                       self._y_ph,
+                                                                                       self.real_sent_lengths_ph)
+        # loss and opt
+        with tf.variable_scope("loss_and_opt"):
+            self.loss = tf.reduce_mean(-log_likelihood)
+            self.train_op = self.get_train_op(self.loss)
+
+        self.sess = tf.Session()
+        self.sess.run(tf.global_variables_initializer())
+        self.load()
+
+    def predict(self, xs):
+        feed_dict = self._fill_feed_dict(xs)
+        logits, trans_params, sent_lengths = self.sess.run([self.logits,
+                                                            self.transition_params,
+                                                            self.real_sent_lengths_ph],
+                                                           feed_dict=feed_dict)
+        # iterate over the sentences because no batching in viterbi_decode
+        y_pred = []
+        for logit, sequence_length in zip(logits, sent_lengths):
+            logit = logit[:int(sequence_length)]  # keep only the valid steps
+            viterbi_seq, viterbi_score = tf.contrib.crf.viterbi_decode(logit, trans_params)
+            y_pred += [viterbi_seq]
+        return y_pred
+
+    def _fill_feed_dict(self, xs, y=None, train=False):
+        assert len(xs) == len(self._xs_ph_list)
+        xs = list(xs)
+        for x in xs[1:]:
+            x = np.array(x)
+        feed_dict = {ph: x for ph, x in zip(self._xs_ph_list, xs)}
+        if y is not None:
+            feed_dict[self._y_ph] = y
+        feed_dict[self.training_ph] = train
+        if not train:
+            feed_dict[self._dropout_ph] = 1.0
+
+        return feed_dict
+
+    def __call__(self, *args, **kwargs):
+        if len(args[0]) == 0 or (args[0] == [0]):
+            return []
+        return self.predict(args)
+
+    def train_on_batch(self, *args):
+        *xs, y = args
+        feed_dict = self._fill_feed_dict(xs, y, train=True)
+        _, loss_value = self.sess.run([self.train_op, self.loss], feed_dict)
+        return {'loss': loss_value,
+                'learning_rate': self.get_learning_rate(),
+                'momentum': self.get_momentum()}
+
+    def load_pretrained_word_emb(self, model_path, model_name, word_dim, word2id=None, vocab_size=None):
+        loaded_words = 0
+        if word2id is not None:
+            vocab_size = len(word2id)
+        word_embeddings = np.zeros(shape=(vocab_size, word_dim))
+
+        if model_name == "glove":
+            model = KeyedVectors.load_word2vec_format(model_path, binary=False)
```

Comment:
Do you really need gensim to load word embeddings?

Reply:
Yes. I think that's a fast and convenient method.